Incorporating Jenny's fixes for the OBO flow

jmprieur opened the pull request:
Incorporating Jenny's fixes for the OBO flow so that the token cache is keyed by the acccess token in the case of a Web API.

Removes a lot of code, uses the recommended patterns
Also some renaming to clarify

See also https://github.com/Azure-Samples/active-directory-dotnet-native-aspnetcore-v2/pull/93

## Purpose
his PR addresses two identified issues:

Token caching was based on the objectId and tenantId of the claims principal generated from the incoming access token.
Token caching, in the case of Web APis / OBO, should be done based on the access token. The current approach does not take into account the access token claims, which could be different for the same user.
For example:
An access token may have MFA or other conditional access policies. This will not be taken into account because the ObjectId and TentantId would still be the same, but the access token now has new claims, which are not verified or taken into consideration.

So, User1 with AT1 -> "{objectIdUser1.tenantId1}" has this cache key for AT1
But, same user completes MFA, so now has a new AT2 w/MFA claims, but will be mapped to same AT1 -> "{objectIdUser1.tenantId1}", so the cache would pull out the wrong access token.

The key for this case should not be "{objectIdUser1.tenantId1}" derived from the incoming access token, but the access token itself.

Call Patterns
On the wiki, it states that we should always use the pattern -> AcquireTokenSilent followed by AcquireTokenInteractive, except for AcquireTokenForClient and AcquireTokenByAuthorizationCode. However, in the sample, a different pattern is used. The pattern is first interactive call (AcquireTokenOnBehalfOf) and then a silent call (AcquireTokenSilent), which is contrary to our guidance and not mentioned in the wiki (that I could find).

This pattern works for OBO because the implementation of AcquireTokenOnBehalfOf does a cache lookup. Which is also never mentioned or documented. It contradicts our guidance.

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]

```

jmprieur commented:
I've tested and this works.
The expiry is taken care of part of the cache configuration.

this PR is ready to review. @MarkZuber : I'm looking forward to getting your feedback on the way the cache key is passed to the token cache serialization logic (I hope it's not too hacky). 
Alternatively would we want to have a .WithCacheKey in msal ?

kalyankrishna1 commented on Microsoft.Identity.Web/HttpContextExtensions.cs:
```diff
@@ -0,0 +1,49 @@
+        /// it can be used in the actions</param>
+        public static void StoreTokenUsedToCallWebAPI(this HttpContext httpContext, JwtSecurityToken token)
+        {
+            httpContext.Items.Add("token", token);
```
>token [](start = 35, length = 5)

"token" is a very generic string and there is a high chance of overwriting a piiece of data that the developer is saving in the HttpContext.Items collection using the key "token". I'd suggest using something hard to guess like "_Msal_cached_access_token_"

TiagoBrenck replied:
I agree that "token" is quite common and it may cause conflicts.

kalyankrishna1 commented on Microsoft.Identity.Web/HttpContextExtensions.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace Microsoft.Identity.Web
+{
+    public static class HttpContextExtensions
```
>HttpContextExtensions [](start = 24, length = 21)

Isn't this JwtBearerOptions.SaveToken() doing the same thing ([https://stackoverflow.com/questions/57057749/what-is-the-purpose-of-jwtbeareroptions-savetoken-property-in-asp-net-core-2-0](https://stackoverflow.com/questions/57057749/what-is-the-purpose-of-jwtbeareroptions-savetoken-property-in-asp-net-core-2-0)) ?

jmprieur replied:
In the code I proposed, we save the decoded JwtSecurityToken, not the raw token.

TiagoBrenck commented on Microsoft.Identity.Web/HttpContextExtensions.cs:
```diff
@@ -0,0 +1,49 @@
+        /// <returns><see cref="JwtSecurityToken"/> used to call the Web API</returns>
+        public static JwtSecurityToken GetTokenUsedToCallWebAPI(this HttpContext httpContext)
+        {
+            return httpContext.Items["token"] as JwtSecurityToken;
```
I have a question about this method. 

In the sample 2, we have a desktop app that calls a web api, that calls graph. There are two ATs involved, web api and graph. Currently, this method is setting only the web api's AT to the `httpContext.Items`. Is that the expected behavior?

jmprieur replied:
yes it is. The token used to call the Web API as we need its hash for the cache key in the case of OBO

TiagoBrenck commented on Microsoft.Identity.Web/TokenCacheProviders/MsalAbstractTokenCacheProvider.cs:
```diff
@@ -69,7 +69,7 @@ private string CacheKey
                 else
                 {
-                    return _httpContextAccessor.HttpContext.User.GetMsalAccountId();
+                    return _httpContextAccessor.HttpContext.GetUserTokenCacheKey();
```
Just a reminder that when we apply this PR to the web-app tutorial, we need to adapt the code to use the new method `SetTokenCacheKey()`.

jmprieur replied:
I discussed with Mark. We need to do better.
cc: @MarkZuber

jmprieur commented on Microsoft.Identity.Web/HttpContextExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+        /// it can be used in the actions</param>
+        public static void StoreTokenUsedToCallWebAPI(this HttpContext httpContext, JwtSecurityToken token)
+        {
+            httpContext.Items.Add("JwtSecurityTokenUsedToCallWebAPI", token);
```
@kalyankrishna1 @TiagoBrenck : PR feedback addressed. renamed "token" to "JwtSecurityTokenUsedToCallWebAPI"

jennyf19 replied:
@jmprieur that name seems a little redundant...what about just "JwtUsedToCallWebAPI"?

jmprieur commented on Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -146,13 +146,21 @@ public async Task AddAccountToCacheFromAuthorizationCodeAsync(AuthorizationCodeR
         /// <param name="tenant">Enables overriding of the tenant/account for the same identity. This is useful in the
         /// cases where a given account is guest in other tenants, and you want to acquire tokens for a specific tenant, like where the user is a guest in</param>
         /// <returns>An access token to call the downstream API and populated with this downstream Api's scopes</returns>
+        /// <remarks>Calling this method from a Web API supposes that you have previously called, 
```
Added pre-requisites

jmprieur commented:
@MarkZuber @kalyankrishna1 @TiagoBrenck : I've addressed your PR feedback and answered your questions. Please re-re-view. 
I think it's now better.

@MarkZuber : the cache key is now computed in the MsalAbstractTokenCacheProvider class, and therefore there is no longer a variable to pass along through the Http Context as it was the case (and as we agreed, was not right)

jmprieur commented:
cc: @jennyf19 and @henrik-me for progress on this

bgavrilMS commented on Microsoft.Identity.Web/TokenCacheProviders/MsalAbstractTokenCacheProvider.cs:
```diff
@@ -69,7 +70,12 @@ private string CacheKey
+                    // should return a token otherwise this might require a challenge
+                    // In the case Web APIs, the token cache key is a hash of the access token used to call the Web API
+                    JwtSecurityToken jwtSecurityToken = _httpContextAccessor.HttpContext.GetTokenUsedToCallWebAPI();
+                    return (jwtSecurityToken != null) ? jwtSecurityToken.RawSignature
```
`jwtSecurityToken.RawSignature` is not a hash of the original access token, as the comment says.

jmprieur replied:
@bgavrilMS : what matters is that the key is different when the JWT is different. Do we agree that we can use the RawSignature as a key for the token cache? even if it's not a hash?
cc: @jennyf19 .

bgavrilMS replied:
Sure

bgavrilMS commented on Microsoft.Identity.Web/TokenAcquisition.cs:
```diff
@@ -162,103 +170,43 @@ public async Task AddAccountToCacheFromAuthorizationCodeAsync(AuthorizationCodeR
+                // Case of Web APIs: we need to do an on-behalf-of flow
+                if (validatedToken != null)
+                {
+                    string tokenUsedToCallTheWebApi = validatedToken.InnerToken == null ? validatedToken.RawData
```
Please add a comment here.

jmprieur replied:
Done. Thanks!

jennyf19 commented:
> using Microsoft.AspNetCore.Http;

do we need license info?

---
Refers to: Microsoft.Identity.Web/HttpContextExtensions.cs:1 in d932a0e. [](commit_id = d932a0e6a27ff8a04c29eea7ceba8c54c30929e0, deletion_comment = False)

jennyf19 commented on Microsoft.Identity.Web/HttpContextExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+    public static class HttpContextExtensions
+    {
+        /// <summary>
+        /// Keep the validated token in associated with the Http request
```
>in associated [](start = 37, length = 13)

remove "in"

kalyankrishna1 commented:
After much discussion, what I learned is that this work needs to be re-looking as the problem we want to solve with it is not a problem really.

_"An access token may have MFA or other conditional access policies. This will not be taken into account because the ObjectId and TentantId would still be the same, but the access token now has new claims, which are not verified or taken into consideration."_  

MSAL Caching implementations will always replace an existing AT in cache when MSAL gets a new one, like a new one with more or less claims than previous ones. If this is not happening then we have a bug in the code. 

_So, User1 with AT1 -> "{objectIdUser1.tenantId1}" has this cache key for AT1_ We just learnt that this is not a good idea after all, as it breaks B2C cases. also it results in cached AT being reused across multiple apps, if the cache is shared and user is the same. We have to be able to set them apart as AT for 1 app will/can be different from another even if the user signing in is the same.

So a new PR in works has the following

1. Using the "sub" claim when an AT is a _user+application_. This works across AAD and B2C and across multiple apps
1. Using the "clientId" when the AT is of type _application_. This is fine as is as a token cache key.

jmprieur commented:
Thanks @kalyankrishna1 for the feedback. I'm not sure for the first point. For the second, yes, awesome. That would moreover work with b2C

Looking forward to your PR. I'd like to merge / fix this by the end of the week.
cc: @jennyf19 @henrik-me FYI

jmprieur commented:
@jennyf19 @henrik-me @bgavrilMS  this is finally merged :) !
